### DISCUSSION TITLE
Docker image?

### DISCUSSION THREAD

**Discussion by murmeltier08** [Thyraz/Sonos-Kids-Controller#12] [Ideas]
I think it could be nice to have a docker fiöe of this project. so everything can run at the server. The raspi show only a website. How diffucult is the creation of this image?

**Comment by NoTrace-Kai** [comment]
Hi. 

I took that image for the raspberry. 
https://github.com/futurice/chilipie-kiosk
And this tutorial
https://www.az-delivery.de/blogs/azdelivery-blog-fur-arduino-und-raspberry-pi/kiosk-modus-fuer-raspberry-pi

perhaps after that you can build a small docker image with the controller software. 

CU Kai

**Comment by murmeltier08** [comment]
This will be my first image. I will try it :-)

**Reply by murmeltier08** [reply]
I still working for the docker file https://github.com/murmeltier08/Sonos-Kids-Controller

I have now the first working docker image. But i have to test it
https://hub.docker.com/repository/docker/murmeltier08/sonos-kids-controller

**Reply by murmeltier08** [reply]
The image work now @Raspberry PI 4

**Comment by murmeltier08** [comment]
@Thyraz  : I can publish the Docker file if you want and you can also refer it to your documentation.  My image run since 1 days without problems

**Comment by murmeltier08** [comment]
@Thyraz Do you would like to have the Dockerfile in your Repo? I would test the Workflow with Docker Hub to your Repo.
What do you think? @stepman0 had this Idea and I find this way a good solution.

**Reply by Thyraz** [reply]
Hi @murmeltier08 and @stepman0,

sorry for being a completely noob about Docker. ;)

A Dockerfile is just text file describing how to build the container right?
Is the container hosted somewhere and what is included?
Is the source (maybe already with compiled www directory) already in there?

Or does it fetch the source from the main repo?

I have just no clue what it means for updates on the main repo.
Needs the dockerfile also to be updated in some cases?
Or just the docker container wherever it is hosted?
(Or nothing at all as long as the dependencies don't change as the source is fetched from the main repo when installing the container?)

**Reply by stepman0** [reply]
Hi @Thyraz, 

a Dockerfile is a recipe for building a container image, which is (don't beat me) comparable to a template of a virtual machine. 

You start with a template (which is the officialy maintained node-image in our case) and add things, copy the source-code from this repo in there an compile everything. Basically the same as your install instructions.

Advantage 1: You can run it on different hardware.
Advantage 2: You don't need to compile the ionic stuff on your raspberry pi - this can be done on a build runner (ok due to qemu for ARM emulation it also takes some time..)

If you look at my fork, I just added the Dockerfile to the repo. The building is done by a Github action and the resulting image is pushed to Dockerhub - which is a registry for docker images.
My fork: https://github.com/stepman0/Sonos-Kids-Controller
Docker registry: https://hub.docker.com/r/stepman0/sonos-kids-controller

One can either pull the pre-build image from this repo or build it on your own machine.

My suggestion would be, that you add the Dockerfile (and if you want the github action for the CI) to your repository. I could help with issues regarding the Dockerfile, but it should be straigtforward.

**Reply by stepman0** [reply]
Btw, this is my Dockerfile: https://github.com/stepman0/Sonos-Kids-Controller/blob/master/Dockerfile

It uses two-stage-building, so the first part is used to compile the ionic stuff, in the second part only the results are copied and therefore the image is smaller (does not contain ionic).

For starting, the one of @murmeltier08 is probably better to understand: https://github.com/murmeltier08/Sonos-Kids-Controller/blob/master/Dockerfile

**Comment by stepman0** [comment]
@murmeltier08 & @Thyraz: I am currently optimizing the Dockerfile (2 stage build to reduce size), I will create a pull request for both of you, as soon as I am finished...

**Comment by murmeltier08** [comment]
@stepman0 Thank you for your nice Input! It is good for me to see how to improve the Dockerfile.
My question is: 
One idea - like @stepman0 one: 
We put the first Dockerfile from @stepman0 Repo in @Thyraz  Repo and I/we try to generate the Docker Image. @stepman0  Do you know the work flow?
 The goal is that we have the Dockerfile in the Main Repo and we can try the image build process. 
If the Dockerfile in the Repo of @Thyraz , did the Image will build automatically or manual? Did he need a docker hub account? 
I think that @Thyraz did not want to maintain the Dockerfile / Image build process If he did not use it?  
@Thyraz what do you think? 

The other solution is that we create a separate Repo for the Dockerfile and reference to @Thyraz one.

**Comment by Thyraz** [comment]
I wrote another noob comment in the pull request:
https://github.com/Thyraz/Sonos-Kids-Controller/pull/16

So still not sure if I understand the whole process and if it's a good idea if I host and maintain this. ;)
At least I should have installed it once on my own, and I should be able to write a documentation about how to install this from the very beginning (naked raspbian os).

I still think a docker container is a very good idea to provide easier access to this project for less tech-savvy users.

**Reply by stepman0** [reply]
I will provide info and readme the next days. I am little busy with homeschooling and stuff...

**Reply by stepman0** [reply]
Finally I found some time to work on this again.

First: I will create a new pull request, as the previous one was based on the fork of murmeltier and I think I messed something with git there - I already made a new fork an committed the Dockerfile. As soon as I added some text to the readme, I will create the PR.

Second: Regarding Docker, I think the best idea is if you create an account on dockerhub (ideally with the same username as on github, makes finding the stuff easier). A free account is sufficient. After that, you create a repository for sonos-kids-controller (only lowercase letters allowed). I guess the name of the repo will then be thyraz/sonos-kids-controller.
On Github, in the settings of this project, you have a section called Secrets. There you click on "New repository secret" (upper right side) and add three secrets, name DOCKER_USERNAME, DOCKER_PASSWORD and DOCKER_REPOSITORY. Username and password of your Dockerhub account, and repository is the name of your repo on dockerhub (likely thyraz/sonos-kids-controller).
The secrets will be used in the Github actions to push the built images to Dockerhub.

I hope this is understandable.

**Comment by murmeltier08** [comment]
@Thyraz Because there is no progress in this topic, you can add perhaps the Repo of @stepman0 in you Documentation? 
Then everyone can use the Docker image from Dockerhub. I add in my Docker hub not only the last version, i have create for every version a new tag

**Reply by stepman0** [reply]
That is possible, the only problem is I have to maintain my fork and keep it updated with Thyraz's. Also I need to maintain the tags in my repo to create versions
